### PR TITLE
distsql: clean the memory usage of MemTracker when a query ends (#10893)

### LINKED ISSUE #10893: Should clean the memory usage of MemTracker when a query ends.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

1. run a select query,
```mysql
select * from t;
```

2. run `SELECT * FROM INFORMATION_SCHEMA.PROCESSLIST;` to show memory usage.
```mysql
mysql> SELECT * FROM INFORMATION_SCHEMA.PROCESSLIST;
+------+------+-----------+------+---------+------+-------+----------------------------------------------+------+
| ID   | USER | HOST      | DB   | COMMAND | TIME | STATE | INFO                                         | MEM  |
+------+------+-----------+------+---------+------+-------+----------------------------------------------+------+
|    2 | root | 127.0.0.1 | test | Sleep   |    1 | 2     |                                              |   96 |
|    1 | root | 127.0.0.1 | test | Query   |    0 | 2     | SELECT * FROM INFORMATION_SCHEMA.PROCESSLIST |    0 |
+------+------+-----------+------+---------+------+-------+----------------------------------------------+------+
2 rows in set (0.00 sec)
```

2. What did you expect to see?

The `Sleep` command is not consuming memory.

3. What did you see instead?

The `Sleep` command is consuming memory.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

closes #10893

### What is changed and how it works?

Correctly release memory consuming in distsql.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

run a select query like `select * from t, ta`

```mysql
mysql> SELECT * FROM INFORMATION_SCHEMA.PROCESSLIST;
+------+------+-----------+------+---------+------+-------+----------------------------------------------+------+
| ID   | USER | HOST      | DB   | COMMAND | TIME | STATE | INFO                                         | MEM  |
+------+------+-----------+------+---------+------+-------+----------------------------------------------+------+
|    1 | root | 127.0.0.1 | test | Sleep   |    3 | 2     |                                              |    0 |
|    2 | root | 127.0.0.1 | test | Query   |    0 | 2     | SELECT * FROM INFORMATION_SCHEMA.PROCESSLIST |    0 |
+------+------+-----------+------+---------+------+-------+----------------------------------------------+------+
2 rows in set (0.00 sec)
```

Code changes

 - None

Side effects

 - Increased code complexity

Related changes

 - None

### CODE REVIEW

File: distsql/select_result.go [hunk 1]
```diff
@@ -106,17 +106,21 @@ func (r *selectResult) fetch(ctx context.Context) {
 			return
```
Comment (crazycs520) [inline distsql/select_result.go]: Should we check in here?

File: distsql/select_result.go [hunk 2]
```diff
@@ -231,16 +234,31 @@ func (r *selectResult) readRowsData(chk *chunk.Chunk) (err error) {
 		}
 	}
 	r.selectResp.Chunks[r.respChkIdx].RowsData = rowsData
+	r.memConsume(-int64(originMemory - r.selectResp.Size()))
```
Comment (crazycs520) [inline distsql/select_result.go]: r.memConsume(-int64(r.selectResp.Size() - originMemory))

File: distsql/distsql_test.go [hunk 3]
```diff
@@ -106,13 +107,30 @@ func (s *testSuite) TestSelectNormal(c *C) {
 	c.Assert(numAllRows, Equals, 2)
 	err := response.Close()
 	c.Assert(err, IsNil)
+	c.Assert(response.memTracker.BytesConsumed(), Equals, int64(0))
+}
+
+func (s *testSuite) TestSelectMemTracker(c *C) {
+
```
Comment (crazycs520) [inline distsql/distsql_test.go]: remove this blank line.

File: distsql/select_result.go (hunk 2 above)
Comment (zz-jason) [reply distsql/select_result.go]: I prefer to directly to use `-`, it indicates a memory releasing.

File: distsql/select_result.go (hunk 1 above)
Comment (SunRunAway) [reply distsql/select_result.go]: Yes, we should. Per https://godoc.org/github.com/pingcap/tidb/kv#Response, resultSubset may be nil even if no error take place.
```go
	// Next returns a resultSubset from a single storage unit.
	// When full result set is returned, nil is returned.
	Next(ctx context.Context) (resultSubset ResultSubset, err error)
```

File: distsql/select_result.go [hunk 4]
```diff
@@ -106,17 +106,21 @@ func (r *selectResult) fetch(ctx context.Context) {
...
-				r.memTracker.Consume(int64(resultSubset.MemSize()))
-			}
+			r.memConsume(int64(resultSubset.MemSize()))
 		}
 
 		select {
 		case r.results <- result:
 		case <-r.closed:
 			// If selectResult called Close() already, make fetch goroutine exit.
+			if resultSubset != nil {
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply distsql/select_result.go]: if line 102 occurs error, `resultSubset` would be `nil`.

File: distsql/select_result.go (hunk 2 above)
Comment (SunRunAway) [reply distsql/select_result.go]: You're right. I'll try to implement in another way.

File: distsql/select_result.go [hunk 5]
```diff
@@ -161,24 +167,21 @@ func (r *selectResult) getSelectResp() error {
...
 		r.selectResp = new(tipb.SelectResponse)
 		err := r.selectResp.Unmarshal(re.result.GetData())
 		if err != nil {
 			return errors.Trace(err)
 		}
-		if r.memTracker != nil && r.selectResp != nil {
-			r.memTracker.Consume(int64(r.selectResp.Size()))
-		}
+		r.selectRespSize = r.selectResp.Size()
+		r.memConsume(int64(r.selectRespSize))
```
Comment (qw4990) [inline distsql/select_result.go]: Why introduce a new variable `selectRespSize` instead of using `r.selectResp.Size()` directly.

Comment (SunRunAway) [reply distsql/select_result.go]: `r.selectResp.Size` is changed because the underlying `Chunks.RowsData` will be cut in `selectResult.readRowsData`.

**Comment by SunRunAway** [comment]
@zz-jason @crazycs520 PTAL again. Thanks.

**Pull request merged**